### ISSUE TITLE
Parser expect immediately rejected Promise return value to be the same as the fn return signature

### ISSUE THREAD

**Issue by NoNameProvided** [Microsoft/TypeScript#10557]
**TypeScript Version:**  1.8.10

**Code**

```ts
test(fail: boolean) : Promise<string> {
    if(fail) {
        return Promise.reject({ any: 'object' });
    }		
    return new Promise( (resolve, reject) => {
        if(Math.random() > 0.5) {
            resolve('my string return value');
        } else { 
            reject({ any: 'object' });
        }
    });
}
```

**Expected behavior:**

The parser should notice I reject the Promise on line 3 and not resolving it, so the return type should not be expected to be the same as the indicated return type. 

Note: It correctly works on line 9 when I reject the promise with the same object signature.

**Actual behavior:** 

The parser returns with the following error for line 3: 

```
Type 'Promise<void>' is not assignable to type 'Promise<string>'. Type 'void' is not assignable to type 'string'.
```

**Comment by NoNameProvided** [comment]
In the meantime I use the `return <any>(Promise.reject( { code: 4 } ));` quick fix, but this way I loose the type, what I don't want.

**Comment by jeffreymorlan** [comment]
This is fixed in TypeScript 2.0, where `Promise.reject` has a return type of `Promise<never>` instead of `Promise<void>`.
